// mlhelper/dist/utils/fileParser/test.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var parser = require("./index");
var path = require("path");
var charts = require("../charts/index");
var index_1 = require("../../algorithm/kNN/index");
var dt = parser.read_csv(path.join(__dirname, '../../../assets/train.csv'), {
    index_col: 0,
    delimiter: ',',
    header: 0,
    dataType: 'number'
});
var labels = dt.getClasses();
var dataSet = dt.drop('quality').values;
var knn = new index_1.default(dataSet, labels);
var dataToTest = parser.read_csv(path.join(__dirname, '../../../assets/test.csv'), {
    index_col: 0,
    dataType: 'number'
}).drop('quality').values;
var resultSet = dataToTest.map(function (v, i) {
    return [i, knn.classify(knn.autoNormalVector(v), 48)];
});
parser.write_csv(path.join(__dirname, './result.csv'), resultSet, {
    header: ['ID', 'quality']
});
var inx = [7.0, 0.27, 0.36, 20.7, 0.045, 45.0, 170.0, 1.001, 3.0, 0.45, 8.8],
    normalInx = knn.autoNormalVector(inx);
console.log(knn.classify(inx, 100));
charts.drawkNN(index_1.default.autoNormal(dataSet), labels, normalInx, {
    width: "500px",
    height: "400px",
    size: 15
});
